include/nlmacros.h

```diff
@@ -92,53 +93,68 @@
+// "a >= 0", in order to avoid "comparison is always true" warnings with
+// unsigned inputs.
+#ifndef ROUNDED_DIV
+#define ROUNDED_DIV(a, b)                       (((a) + (((a) == 0 || (a) > 0) != ((b) == 0 || (b) > 0) ? -(b) : (b)) / 2) / (b))
+#endif
+
 #else // #ifdef BUILD_FEATURE_USE_UNSAFE_MACROS
 
 // Safer macros.
 
 // MIN and MAX include static asserts that trigger if their two arguments
-// differ in signedness.  This prevents, for example, MAX(-1,1U) = -1.
+// differ in signedness. This prevents, for example, MAX(-1,1U) = -1.
 //
 // The test for each argument is "(__typeof__(X))(-1) <= 0"; casting the "-1"
 // to the type of X forces it to become unsigned if X is unsigned, so the test
 // simplifies to either "-1 <= 0" (true) or "-1U <= 0" (false, because "-1U"
-// is a large all-ones positive number).  If the two macro arguments have
+// is a large all-ones positive number). If the two macro arguments have
 // different signedness, the two test results will be different and we'll
 // trigger the assert.
 //
 // Note that the comparison is "<= 0" rather than "< 0" in order to avoid
 // "comparison is always false" warnings when X is unsigned.
 //
 // Note also that we're testing the signedness AS WRITTEN (that is, before
-// integer promotion).  This means that we will not assert on a comparison
+// integer promotion). This means that we will not assert on a comparison
 // between uint16_t and uint32_t, and we WILL assert on a comparison between
 // uint16_t and int32_t, even though in both cases the compiler will internally
 // promote the unsigned uint16_t to a SIGNED int_32_t for the comparison.
+//
+// Finally, note that the LOG2 and CTZ macros explicitly cast their parameters
+// to unsigned int, because that's the argument type of the underlying GCC
+// builtin functions.
 #ifndef MIN
 #define MIN(a, b)                               MIN_INNER(a, b, __COUNTER__)
 #define MIN_INNER(a, b, C)                      ({ __typeof__(a) _CC(_a, C) = (a); __typeof__(b) _CC(_b, C) = (b); _NLMACROS_ASSERT(((__typeof__(a))(-1) <= 0) == ((__typeof__(b))(-1) <= 0)); (_CC(_a, C) < _CC(_b, C)) ? _CC(_a, C) : _CC(_b, C); })
 #endif
-
 #ifndef MAX
 #define MAX(a, b)                               MAX_INNER(a, b, __COUNTER__)
 #define MAX_INNER(a, b, C)                      ({ __typeof__(a) _CC(_a, C) = (a); __typeof__(b) _CC(_b, C) = (b); _NLMACROS_ASSERT(((__typeof__(a))(-1) <= 0) == ((__typeof__(b))(-1) <= 0)); (_CC(_a, C) > _CC(_b, C)) ? _CC(_a, C) : _CC(_b, C); })
 #endif
-
+#ifndef LIMIT
+#define LIMIT(v, low, high)                     MAX(low, MIN(v, high))
+#endif
 #ifndef LOG2
 #define LOG2(x)                                 LOG2_INNER(x, __COUNTER__)
-#define LOG2_INNER(x, C)                        ({ __typeof__(x) _CC(_x, C) = (x); (_CC(_x, C) != 0) ? 31 - __builtin_clz(_CC(_x, C)) : -1; })
+#define LOG2_INNER(x, C)                        ({ unsigned int _CC(_x, C) = (x); (_CC(_x, C) != 0) ? 31 - __builtin_clz(_CC(_x, C)) : -1; })
```

Comment:
This change and the one below on 157 seem unrelated. What was the genesis of these? Regardless, they should probably be in a separate PR if they are unrelated to the addition of LIMIT, ROUNDED_DIV, and SIZEOF.

Reply:
You're right, these two changes are unrelated to the others.  And they're unfinished, too.  I'll pull them out into a separate PR.